File: apple/LC45JumpGameII.py
class Solution(object):
    def jump(self, nums):
        if not nums:
            return 0
        start = 0
        end = 0
        step = 0
        while end < len(nums)-1:
            step += 1
            currentmax = end+1
            for i in range(start, currentmax):
                if i + nums[i] >= len(nums)-1:
                    return step
                currentmax = max(currentmax, i+nums[i])
            start, end = end, currentmax
        return step

File: apple/LC45JumpGameII.java
class Solution {
    public int jump(int[] nums) {
        if (nums.length == 0 || nums == null) return 0;
        int start = 0;
        int end = 0;
        int answer = 0;
        while (end < nums.length-1) {
            answer++;
            int currentend = end+1;
            for (int i = start; i < end+1; i++) {
                if (i + nums[i] >= nums.length -1) return answer;
                currentend = Math.max(currentend, i+nums[i]);
            }
            start = end;
            end = currentend;
        }
        return answer;
    }
}
